File: MPI.py
import numpy as np
from mpi4py import MPI

def streams(total, SIZE, RANK):
    n = total
    q = n // SIZE
    r = SIZE * q - n

    chunk = q
    if (RANK >= SIZE - r):
        chunk = q - 1
    return chunk

SIZE = MPI.COMM_WORLD.Get_size()
RANK = MPI.COMM_WORLD.Get_rank()
NAME = MPI.Get_processor_name()

n = 100
t = MPI.Wtime()
nrows = streams(n, SIZE, RANK)
#a = [None]*nrows*(n+1)
#x = [None]*n
#tmp = [None]*(n+1)
tmp = np.ones(n+1)
a = np.ones(nrows*(n+1))
x = np.ones(n)
#x = np.arange (0, n, 1)
#x = range (0, n)
rows = np.ones(nrows+1)
#rows = [None]*nrows
for i in range(0, nrows):
    rows[i] = RANK + SIZE * i
    for j in range(0, n-1):
        a[i*(n-1)+j] = np.random.randint(1, 99)
    a[i*(n+1)+n] = np.random.randint(1, 99)

#if RANK == 0: print(rows)

row = 0
for i in range(0, n-1):
    if i == rows[row]:
        MPI.COMM_WORLD.Bcast (a[row * (n+1)], root = RANK)
        for j in range(0, n+1):
            tmp[j]=a[row*(n+1)+j]
        row = row + 1
    else:
        MPI.COMM_WORLD.Bcast (tmp, root = i // SIZE)
    for j in range(row, nrows):
        scaling = a[j*(n+1) + i] / tmp[i]
        for k in range(i, n+1):
            a[j * (n + 1) + k] = a[j*(n+1)+k] - scaling * tmp[k]

row = 0
for i in range(0, n):
    x[i] = 0
    if (i == rows[row]):
        x[i] = a[row*(n+1) + n]
        row = row + 1

row = nrows - 1
i = n-1
while i > 0:
    if (row >= 0):
        if (i == rows[row]):
            x[i] = x[i] / a[row * (n+1) + i]
            MPI.COMM_WORLD.Bcast (x[i], root = RANK)
            row = row - 1
        else:
            MPI.COMM_WORLD.Bcast (x[i], root = i // SIZE)
    else:
        MPI.COMM_WORLD.Bcast(x[i], root = i // SIZE)
    for j in range(0, row+1):
        x[rows[j]] = x[rows[j]] - a[j * (n+1) + i] * x[i]
    i = i - 1

if (RANK == 0):
    x[0] = x[0] / a[row * (n+1)]
MPI.COMM_WORLD.Bcast (x, root=0)

t = MPI.Wtime - t
if (RANK == 0):
    print("Time: ", t)

File: mpi.c
#include <math.h>
#include <mpi.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
int streams(int total, int commsize, int rank)
{
    int n = total;
    int q = n / commsize;
    if (n % commsize != 0) q++;
    int r = commsize * q - n;

    int chunk = q;
    if (rank >= commsize - r) chunk = q - 1;
    return chunk;
}

int main(int argc, char *argv[])
{
    int n = 1000;
    int rank, commsize;
    MPI_Init(&argc,&argv);
    double t = MPI_Wtime();
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &commsize);

    int nrows = streams(n, commsize, rank);
    int *rows = malloc(sizeof(*rows) * nrows);
printf("%d, %d, %d\n", commsize, rank, nrows);
    double *a = malloc(sizeof(*a) * nrows * (n+1));
    double *x = malloc(sizeof(*x) * n);
    double *tmp = malloc(sizeof(*tmp) * (n+1));

    //Generate

    for (int i=0; i < nrows; i++)
    {
        rows[i] = rank + commsize * i;
        srand(time(NULL));
        for (int j=0; j<n; j++)
            a[i*(n+1)+j] = rand() % 100 + 1;
        //b
        a[i*(n+1)+n] = rand() % 100 + 1;
    }
    //Прямой
    int row = 0;
    for (int i=0; i<n-1; i++)
    {
        //Исключаем xi
        if (i == rows[row])
        {
            //Рассылка строки i
            MPI_Bcast(&a[row * (n+1)], n+1, MPI_DOUBLE, rank, MPI_COMM_WORLD);
            for (int j=0; j<=n; j++)
                tmp[j]=a[row * (n+1) + j];
            row++;
        }
        else
        {
            MPI_Bcast(tmp, n+1, MPI_DOUBLE, i % commsize, MPI_COMM_WORLD);
        }
        //Вычет строки
        for (int j=row; j<nrows; j++)
        {
            double scaling = a[j*(n+1) + i] / tmp[i];
            for (int k=i; k<n+1; k++)
                a[j*(n+1)+k] -= scaling * tmp[k];
        }
    }

    //"Грязная" инициализация x
    row = 0;
    for (int i=0; i<n; i++)
    {
        x[i]=0;
        if (i==rows[row])
        {
            x[i] = a[row*(n+1) + n];
            row++;
        }
    }

    //Обратный
    row = nrows-1;
    for (int i=n-1; i>0; i--)
    {
        if (row >= 0)
        {
            if (i == rows[row])
            {
                x[i] /= a[row * (n+1) +i];
                MPI_Bcast(&x[i], 1, MPI_DOUBLE, rank, MPI_COMM_WORLD);
                row--;
            }
            }
            else MPI_Bcast(&x[i], 1, MPI_DOUBLE, i%commsize, MPI_COMM_WORLD);
        }
        else MPI_Bcast(&x[i], 1, MPI_DOUBLE, i%commsize, MPI_COMM_WORLD);

        for (int j=0; j<=row; j++)
            x[rows[j]] -= a[j * (n+1) + i] * x[i];
    }

    if (rank == 0)
        x[0] /= a[row * (n+1)];
    MPI_Bcast(x, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);

    t = MPI_Wtime() - t;

    if (rank == 0)
        printf("Gauss MPI: n %d, procs %d, time (sec) %.6f\n", n, commsize, t);
        //printf("%d,%d,%d,%d,%d,%d \n", x[0], x[1], x[2], x[3], x[4], x[5]);
    MPI_Finalize();
    return 0;
}

